File: tools/nasa_gif.py
#!/usr/bin/env python3
"""
nasa_gif.py — Python version of the NASA image terminal tool
Bonus: renders images as ASCII art in terminal if `chafa` is installed.

Usage:
  python3 nasa_gif.py search <query>
  python3 nasa_gif.py apod [YYYY-MM-DD]
  python3 nasa_gif.py random
  python3 nasa_gif.py download <url> <filename>
  python3 nasa_gif.py ascii <url>        # Render image as terminal ASCII

Requirements: Python 3.8+, no third-party packages required.
Optional: brew install chafa  (for ASCII art rendering)
"""

import sys
import os
import json
import urllib.request
import urllib.parse
import subprocess
import shutil
import random as rand_module
from datetime import date, timedelta

NASA_API_KEY = os.environ.get('NASA_API_KEY', 'DEMO_KEY')
NASA_IMAGES_BASE = 'https://images-api.nasa.gov'
NASA_APOD_BASE = f'https://api.nasa.gov/planetary/apod?api_key={NASA_API_KEY}'
DOWNLOAD_DIR = os.path.join(os.path.dirname(__file__), 'downloads')

# ── helpers ────────────────────────────────────────────────────────────────────

def fetch_json(url):
    req = urllib.request.Request(url, headers={'User-Agent': 'lobit-hydra/1.0'})
    with urllib.request.urlopen(req, timeout=15) as resp:
        return json.loads(resp.read().decode())

def divider():
    print('─' * 60)

def print_item(item, idx):
    data = item.get('data', [{}])[0]
    links = item.get('links', [])
    preview = next((l['href'] for l in links if l.get('rel') == 'preview'),
                   links[0]['href'] if links else '(no preview)')
    divider()
    print(f"[{idx+1}] {data.get('title', 'Untitled')}")
    print(f"    Type: {data.get('media_type','image')} | Date: {str(data.get('date_created',''))[:10]}")
    desc = (data.get('description') or '').replace('\n', ' ')[:120]
    print(f"    {desc}...")
    print(f"    Preview: {preview}")
    print(f"    NASA ID: {data.get('nasa_id', 'n/a')}")

def download_file(url, dest):
    os.makedirs(DOWNLOAD_DIR, exist_ok=True)
    path = os.path.join(DOWNLOAD_DIR, dest)
    print(f"Downloading to {path}...")
    urllib.request.urlretrieve(url, path)
    print(f"Saved: {path}")
    return path

# ── commands ───────────────────────────────────────────────────────────────────

def cmd_search(query, media_type=''):
    type_param = f'&media_type={media_type}' if media_type else ''
    url = f"{NASA_IMAGES_BASE}/search?q={urllib.parse.quote(query)}{type_param}&page_size=12"
    print(f'\nSearching NASA library for: "{query}"\n')
    data = fetch_json(url)
    items = data.get('collection', {}).get('items', [])
    if not items:
        print('No results found.')
        return
    for idx, item in enumerate(items):
        print_item(item, idx)
    divider()
    print(f'\n{len(items)} results. Download: python3 nasa_gif.py download <url> <filename>')

def cmd_apod(date_str=None):
    date_param = f'&date={date_str}' if date_str else ''
    url = f'{NASA_APOD_BASE}{date_param}'
    label = f'for {date_str}' if date_str else '(today)'
    print(f'\nFetching APOD {label}...\n')
    data = fetch_json(url)
    divider()
    print(f"Title: {data.get('title')}")
    print(f"Date:  {data.get('date')}")
    print(f"Type:  {data.get('media_type')}")
    asset_url = data.get('hdurl') or data.get('url', '')
    print(f"URL:   {asset_url}")
    if data.get('thumbnail_url'):
        print(f"Thumb: {data['thumbnail_url']}")
    divider()
    print(f"\n{str(data.get('explanation',''))[:300]}...\n")
    ext = asset_url.split('.')[-1].split('?')[0]
    print(f"Download: python3 nasa_gif.py download \"{asset_url}\" \"apod_{data.get('date')}.{ext}\"")

def cmd_random():
    start = date(1995, 6, 16)
    end = date.today() - timedelta(days=1)
    delta = (end - start).days
    rand_date = start + timedelta(days=rand_module.randint(0, delta))
    print(f'Random APOD date: {rand_date}')
    cmd_apod(str(rand_date))

def cmd_download(url, filename):
    if not url or not filename:
        print('Usage: python3 nasa_gif.py download <url> <filename>')
        sys.exit(1)
    download_file(url, filename)

def cmd_ascii(url):
    """Download image and render as ASCII art using chafa if available."""
    if not shutil.which('chafa'):
        print('Install chafa for ASCII rendering: brew install chafa')
        print(f'URL: {url}')
        return
    tmp = '/tmp/lobit_nasa_preview.jpg'
    urllib.request.urlretrieve(url, tmp)
    subprocess.run(['chafa', '--size', '80x40', tmp])

def cmd_gif(query):
    print('NASA Images API does not index GIFs directly.')
    print('Searching for videos and animated content...\n')
    cmd_search(query, media_type='video')

# ── main ───────────────────────────────────────────────────────────────────────

def usage():
    print("""
lobit-hydra :: NASA image tool (Python)

  python3 nasa_gif.py search <query>
  python3 nasa_gif.py gif <query>          Search videos/animations
  python3 nasa_gif.py apod                 Today's APOD
  python3 nasa_gif.py apod <YYYY-MM-DD>    APOD for a date
  python3 nasa_gif.py random               Random APOD
  python3 nasa_gif.py download <url> <f>   Download to ./downloads/
  python3 nasa_gif.py ascii <url>          Render in terminal (needs chafa)

Set NASA_API_KEY env var for higher rate limits.
""")

if __name__ == '__main__':
    args = sys.argv[1:]
    if not args:
        usage()
        sys.exit(0)

    cmd = args[0]
    rest = args[1:]

    if cmd == 'search':
        cmd_search(' '.join(rest) or 'nebula')
    elif cmd in ('gif', 'gifs'):
        cmd_gif(' '.join(rest) or 'space')
    elif cmd == 'apod':
        cmd_apod(rest[0] if rest else None)
    elif cmd == 'random':
        cmd_random()
    elif cmd == 'download':
        cmd_download(rest[0] if rest else None, rest[1] if len(rest) > 1 else None)
    elif cmd == 'ascii':
        cmd_ascii(rest[0] if rest else '')
    else:
        usage()

File: tools/nasa_gif.js
#!/usr/bin/env node
// nasa_gif.js — Terminal tool for finding NASA imagery for live event use
// Downloads are rendered as MP4 and saved to the project library.
//
// Usage:
//   node nasa_gif.js search <query>           Search NASA image/video library
//   node nasa_gif.js apod                     Get today's Astronomy Picture of the Day
//   node nasa_gif.js apod <YYYY-MM-DD>        Get APOD for specific date
//   node nasa_gif.js download <url> <name>    Download + render to library as <name>.mp4
//   node nasa_gif.js random                   Random space image (great for VJ drops)
//   node nasa_gif.js list                     Show all videos in the library
//
// API key: uses DEMO_KEY (1000 req/day, 30/hr). Set NASA_API_KEY env var for more.

import { createWriteStream, existsSync, mkdirSync, rmSync } from 'fs'
import { pipeline } from 'stream/promises'
import https from 'https'
import http from 'http'
import path from 'path'
import { tmpdir } from 'os'
import { fileURLToPath } from 'url'
import {
  buildSearchUrl,
  buildApodUrl,
  randomApodDate,
  parseSearchItems,
  extractApodAssetUrl,
  formatApodFilename,
  extractPreviewUrl,
} from './nasa-core.js'
import { saveToLibrary, readIndex } from './library.js'

const __dirname = path.dirname(fileURLToPath(import.meta.url))
const NASA_API_KEY = process.env.NASA_API_KEY || 'DEMO_KEY'

// ── helpers ───────────────────────────────────────────────────────────────────

function fetchJSON(url) {
  return new Promise((resolve, reject) => {
    const client = url.startsWith('https') ? https : http
    client.get(url, (res) => {
      let data = ''
      res.on('data', (chunk) => data += chunk)
      res.on('end', () => {
        try { resolve(JSON.parse(data)) }
        catch (e) { reject(new Error(`JSON parse failed: ${e.message}\n${data.slice(0, 200)}`)) }
      })
    }).on('error', reject)
  })
}

function downloadToPath(url, destPath) {
  const dir = path.dirname(destPath)
  if (!existsSync(dir)) mkdirSync(dir, { recursive: true })
  return new Promise((resolve, reject) => {
    const client = url.startsWith('https') ? https : http
    client.get(url, (res) => {
      if (res.statusCode === 301 || res.statusCode === 302) {
        return downloadToPath(res.headers.location, destPath).then(resolve).catch(reject)
      }
      if (res.statusCode !== 200) return reject(new Error(`HTTP ${res.statusCode}`))
      const writer = createWriteStream(destPath)
      pipeline(res, writer).then(() => resolve(destPath)).catch(reject)
    }).on('error', reject)
  })
}

function printDivider() { console.log('─'.repeat(60)) }

function printItem(item, idx) {
  const data = item.data?.[0] ?? {}
  const mediaType = data.media_type || 'image'
  const preview = extractPreviewUrl(item) ?? '(no preview)'

  printDivider()
  console.log(`[${idx + 1}] ${data.title || 'Untitled'}`)
  console.log(`    Type: ${mediaType} | Date: ${data.date_created?.slice(0, 10) || 'unknown'}`)
  console.log(`    ${(data.description || '').slice(0, 120).replace(/\n/g, ' ')}...`)
  console.log(`    Preview: ${preview}`)
  console.log(`    NASA ID: ${data.nasa_id || 'n/a'}`)
}

// ── commands ──────────────────────────────────────────────────────────────────

async function search(query, mediaType = '') {
  const url = buildSearchUrl(query, mediaType)
  console.log(`\nSearching NASA library for: "${query}"\n`)

  const data = await fetchJSON(url)
  const items = parseSearchItems(data)

  if (!items.length) { console.log('No results found.'); return }

  items.forEach((item, idx) => printItem(item, idx))
  printDivider()
  console.log(`\n${items.length} results.`)
  console.log(`Add to library: node nasa_gif.js download <url> <name>`)
}

async function apod(dateStr) {
  const url = buildApodUrl(dateStr, NASA_API_KEY)
  console.log(`\nFetching APOD${dateStr ? ` for ${dateStr}` : ' (today)'}...\n`)

  const data = await fetchJSON(url)
  const assetUrl = extractApodAssetUrl(data)
  const suggestedName = formatApodFilename(data).replace(/\.[^.]+$/, '')

  printDivider()
  console.log(`Title: ${data.title}`)
  console.log(`Date:  ${data.date}`)
  console.log(`Type:  ${data.media_type}`)
  console.log(`URL:   ${assetUrl}`)
  if (data.thumbnail_url) console.log(`Thumb: ${data.thumbnail_url}`)
  printDivider()
  console.log(`\n${data.explanation?.slice(0, 300)}...\n`)
  console.log(`Add to library: node nasa_gif.js download "${assetUrl}" "${suggestedName}"`)
}

async function random() {
  const dateStr = randomApodDate()
  console.log(`Random APOD date: ${dateStr}`)
  await apod(dateStr)
}

async function download(url, name) {
  if (!url || !name) {
    console.error('Usage: node nasa_gif.js download <url> <name>')
    console.error('  name: library entry name (no extension) — used in initVideo()')
    process.exit(1)
  }

  const rawExt = path.extname(url.split('?')[0]).toLowerCase() || '.jpg'
  const tmpPath = path.join(tmpdir(), `lobit_${Date.now()}${rawExt}`)

  console.log(`\nDownloading...`)
  await downloadToPath(url, tmpPath)

  console.log(`Rendering as MP4 → library...`)
  const dest = await saveToLibrary(tmpPath, name, { source: url })
  rmSync(tmpPath, { force: true })

  console.log(`\nSaved: ${dest}`)
  console.log(`Hydra: initVideo('${name}')`)
}

async function list() {
  const entries = readIndex()
  if (!entries.length) {
    console.log('\nLibrary is empty.')
    console.log('Add a video: node nasa_gif.js download <url> <name>\n')
    return
  }

  printDivider()
  entries.forEach((e, i) => {
    const name = e.filename.replace(/\.mp4$/, '')
    console.log(`[${i}] ${name}`)
    if (e.added) console.log(`    Added: ${e.added.slice(0, 10)}`)
    if (e.source) console.log(`    ${e.source.slice(0, 72)}`)
  })
  printDivider()
  console.log(`\n${entries.length} video${entries.length === 1 ? '' : 's'} in library`)
  console.log(`Hydra: initVideo('name')  — see tools/init_video.js\n`)
}

// ── main ──────────────────────────────────────────────────────────────────────

const [,, cmd, ...args] = process.argv

switch (cmd) {
  case 'search':
    await search(args.join(' ') || 'nebula')
    break
  case 'apod':
    await apod(args[0])
    break
  case 'random':
    await random()
    break
  case 'download':
    await download(args[0], args[1])
    break
  case 'list':
    await list()
    break
  case 'gifs':
  case 'gif':
    await search(args.join(' ') || 'space', 'video')
    break
  default:
    console.log(`
lobit-hydra :: NASA image tool

  node nasa_gif.js search <query>           Search NASA image library
  node nasa_gif.js gif <query>              Search NASA video/animation library
  node nasa_gif.js apod                     Today's Astronomy Picture of the Day
  node nasa_gif.js apod <YYYY-MM-DD>        APOD for a specific date
  node nasa_gif.js random                   Random space image (random APOD)
  node nasa_gif.js download <url> <name>    Download + render to library as <name>.mp4
  node nasa_gif.js list                     List all videos in the library

Set NASA_API_KEY env var to use your own key (default: DEMO_KEY, 30 req/hr).
Start library server:  npm run library:serve  → http://localhost:3001
Hydra usage:           initVideo('name')      — see tools/init_video.js
`)
}
